firstapp/app/src/androidTest/java/com/example/firstapp/ShellUtility.java

```diff
@@ -138,55 +158,123 @@ public static UiObject castToObject(UiObject2 object2) throws invalidInputExcept
+                break;
+
             case "click":
-                if (strict) {
-                    object2 = device.wait(Until.findObject(By.text(text)), timeoutMs);
-                } else {
-                    object2 = device.wait(Until.findObject(By.textContains(text)), timeoutMs);
-                }
-                object = castToObject(object2);
-                object.click();
-                return object;
+                action = new ClickAction();
+                ((ClickAction) action).text = tokens[1];
+                break;
 
             case "clickImage":
-                if (strict) {
-                    object2 = device.wait(Until.findObject(By.desc(text)), timeoutMs);
-                } else {
-                    object2 = device.wait(Until.findObject(By.descContains(text)), timeoutMs);
-                }
-                object = castToObject(object2);
-                object.click();
-                return object;
+                action = new ClickImageAction();
+                ((ClickImageAction) action).description = tokens[1];
+                break;
 
             case "edit":
-                String entered = tokens[2];
-                if (strict) {
-                    object2 = device.wait(Until.findObject(By.text(text)), timeoutMs);
-                } else {
-                    object2 = device.wait(Until.findObject(By.textContains(text)), timeoutMs);
-                }
-                object = getEditableObject(object2);
-                object.waitForExists(timeoutMs);
-                object.legacySetText(entered);
-                return object;
+                action = new EditAction();
+                ((EditAction) action).text = tokens[1];
+                ((EditAction) action).entered = tokens[2];
+                break;
 
             default:
-                throw new invalidInputException("an action has an invalid first token.");
+                throw new invalidInputException("input at index " + i + " has an invalid first token");
         }
+        action.strict = tokens[tokens.length - 1].equals("strict");
+        return action;
     }
 
+    public Action[] parseStringCUJ(String[] CUJ) throws invalidInputException {
+        Action[] res = new Action[CUJ.length];
+        for (int i = 0; i < res.length; i++) {
+            res[i] = parseStringAction(CUJ[i], i);
+        }
+        return res;
+    }
+
+    /**
+     * Executes an action with parts speficied by tokens and returns a UiObject (for caching)
+     */
+    public UiObject executeUncachedAction(Action action, int i) throws invalidInputException, UiObjectNotFoundException, IOException, InterruptedException {
+        boolean strict = action.strict;
+
+        UiObject2 object2 = null;
+        UiObject object = null;
+        if (action instanceof StartAction) {
+            String pkg = ((StartAction) action).pkg;
+            forceQuitApp(pkg);
+            launchApp(pkg);
```

Comment:
This code can be moved into the action itself.  Have the action be responsible for executing the code to fulfill the action.  Then you can unit test each individual Action:

    StartActionTest
    ClickActionTest
    ClickImageActionTest
    ...

Or make Action be an enum and leave `executeUncachedAction()` as is.

Reply:
Actually, the enum wouldn't work because Action is also storing data specific to an action.